sycl/source/detail/scheduler/graph_builder.cpp

```diff
@@ -355,10 +355,9 @@ Command *Scheduler::GraphBuilder::insertMemoryMove(MemObjRecord *Record,
 
     // Full copy of buffer is needed to avoid loss of data that may be caused
     // by copying specific range from host to device and backwards.
-    NewCmd =
-        new MemCpyCommand(*AllocaCmdSrc->getRequirement(), AllocaCmdSrc,
-                          *AllocaCmdDst->getRequirement(), AllocaCmdDst,
-                          AllocaCmdSrc->getQueue(), AllocaCmdDst->getQueue());
+    NewCmd = new MemCpyCommand(*AllocaCmdSrc->getRequirement(), AllocaCmdSrc,
+                               *Req, AllocaCmdDst, AllocaCmdSrc->getQueue(),
+                               AllocaCmdDst->getQueue());
   }
```

Comment:
@sergey-semenov ,   

If we do not have unified memory, then when moving back to a host accessor we enqueue both a Read and a Write operation.  If the host accessor uses the `discard_write` access mode, then the Read operation is unnecessary. In this case, there is a big performance improvement if the Read operation is not scheduled.

As you know, we already have the check for that elsewhere.   However, that check is not working. And the reason it's not working is that here the `Req` argument has the `discard_write` access mode.  But the `AllocaCmdDst` does not.

So we have a situation where there isn't a clear single source for truth.  Now, as I've been going through this, it seems like this fix is the right approach.  That we should use the `Req` argument, not `AllocaCmdDst->getRequirement()`.

I know this change results in a test failure, but ignoring that for a moment, does this fix seem like the  correct approach to you?

If this approach is fine, I will continue to address the test failure, which is due to an unrelated bug that we've now unearthed. But if this approach isn't optimal, let me know and I will take a different approach.

Reply:
I think if that check in `MemCpyCommand::enqueueImp` does not work anyway, we might as well apply the TODO mentioned there and not generate the command at all in that case to save us the creation/cleanup overhead. I've tried a rough version of such a change locally, and it doesn't seem to cause any problems (at least with the in-tree tests), so I would recommend looking into that approach.